### PR TITLE
Optimize ByteCountCoded<> to not do extra heap allocations

### PR DESCRIPTION
**Pull request by ZappoMan**
Adds support for ByteCountCoded<> to operate against an existing buffer as well.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: c0cdf3256cca585fdc6f5d72304f45faf22c143f available at: https://gist.github.com/90c1a952cab570b42456<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5283.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5283.dmg

File: libraries/shared/src/ByteCountCoding.h [hunk 1]
```diff
@@ -111,51 +116,56 @@ template<typename T> inline QByteArray ByteCountCoded<T>::encode() const {
...
 
 template<typename T> inline void ByteCountCoded<T>::decode(const QByteArray& fromEncodedBytes) {
+    decode(fromEncodedBytes.constData(), fromEncodedBytes.size());
+}
 
-    // first convert the ByteArray into a BitArray...
-    QBitArray encodedBits;
-    int bitCount = BITS_IN_BYTE * fromEncodedBytes.count();
-    encodedBits.resize(bitCount);
+template<typename T> inline void ByteCountCoded<T>::decode(const char* encodedBuffer, int encodedSize) {
```
Comment (jherico) [inline libraries/shared/src/ByteCountCoding.h]: this function should return a size_t indicating how many bytes from the encoded buffer were consumed as part of the decode process.

Comment (ZappoMan) [inline libraries/shared/src/ByteCountCoding.h]: excellent suggestion. will do

Comment (ZappoMan) [inline libraries/shared/src/ByteCountCoding.h]: don't merge as is, this request helped me notice another optimization for large buffers

**Comment by hifi-gustavo** [comment]
Build console output for commit: 7790e35d2497ca6ff631413ba943f9bf8ea63b3d available at: https://gist.github.com/90c1a952cab570b42456<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5283.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5283.dmg

**Pull request merged**